transaction/core/src/ring_signature/rct_bulletproofs.rs

```diff
@@ -42,9 +65,29 @@ pub struct SignatureRctBulletproofs {
     /// This contains range_proof.to_bytes(). It is stored this way so that this
     /// struct may derive Default, which is a requirement for serializing
     /// with Prost.
+    ///
+    /// Note: This is EMPTY if mixed transactions are enabled
     #[prost(bytes, tag = "3")]
     #[digestible(never_omit)]
     pub range_proof_bytes: Vec<u8>,
+
+    /// A range proof, one for each token id that is used in the transaction.
+    ///
+    /// The range proofs correspond to the sorted order of token ids used.
+    ///
+    /// Note: This is EMPTY if mixed transactions is not enabled
+    #[prost(bytes, repeated, tag = "4")]
```

Comment:
Similarly to the oneof, I think we could use an enum type: RangeProof {
SingleProof: Vec<u8>
MultiProof: Vec<Vec<u8>>
}

Reply:
So the way I think the schema evolution is likely to work is:

Initially: Only `Vec<u8>`

Transition: Both a `Vec<u8>` field and a `Vec<Vec<u8>>` field

End: Remove the (no longer used) `Vec<u8>` field, and only the `Vec<Vec<u8>>` field remains.

This is normally the forwards-and-backwards compatibilie method of protobuf schema evolution. If we already had a OneOf here then we could use that method instead, but we can't just change the `Vec<u8>` to a `OneOf` now without a breaking change.

We could make the next thing `OneOf<Vec<Vec<u8>>>` instead. I'm not sure if that makes sense. If Isis is right then we may be able to aggregate all these range proofs. But even in that case `Vec<Vec<u8>>` is fine.